docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+# Sync improvements
+
+* Author(s): Cornelius Weig (@corneliusweig)
+* Design Shepherd: Tejal Desai (@tejal29)
+* Date: 03/20/2019
+* Status: Draft
+
+## Background
+
+Currently, skaffold config supports `artifact.sync` as a way to sync files
+directly to pods. So far, artifact sync requires a specification of sync
+patterns like
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
+In addition, the syncing needs to handle special cases for globbing and often
+requires a long list of sync patterns (#1807).
+
+Furthermore, builders should be able to inform Skaffold about additional sync paths (#1704).
+This will result in much faster deploy cycles, without users having to bother about these optimizations.
+
+## Design
+
```

Comment:
I'd add a section here about 3 sync modes: 

   - "smart": builder recommended sync mapping for files, e.g. jib gives you both the source and dest of files to sync, you don't have to spec anything else 
   - "auto-inferred destination": you specify the files you're interested in, we infer the destination in the container, e.g based on Dockerfile 
   - "direct": you specify both the source and destination

I don't want to combine these three modes into hybrid scenarios _at first_. We could just have them implemented separately and figure out the combination rules later, if people are asking for it! 

"smart" is only available for jib (maybe?) and it is itself a separate design, so maybe we should just mention it here that the details are out of scope for this design proposal, but it should fit in the model as one of the options users can choose.

Reply:
About the hybrid scenario: "direct" and "auto" mode can easily be combined. Is there any reason why you want to avoid this in particular?

Reply:
My guess is that: 

- For non-docker users it's only going to be "manual". 
- For Dockerfile users 99% of the cases should be covered by "infer". 
- For jib users hopefully most of it will be covered by "smart"...

Thus, I did not want to deal with the complexity of combining the "manual" and "infer" scenarios, unless it's really required. Why maintain precedence rules and merging logic code if no one will use it?

Reply:
this is not saying that _when_ the above assumptions fail and users are asking for hybrid features we can't implement them - we can, but then we'll have some more data to work with.

Reply:
Alright, you convinced me :)

Reply:
It seems to me that the Docker builder can be "smart" too: just apply the inferred rules *providing* there are no `RUN` statements after an `ADD` or `COPY`.  I don't know how common this case would be in practice though.

Reply:
That's a funny idea. I'll add that as possible future work.